src/rtc_pl031.rs

```diff
@@ -0,0 +1,612 @@
+/// let delay = Duration::from_millis(1500);
+/// thread::sleep(delay);
+///
+/// // Read the current RTC value from the Data Register
+/// rtc.read(RTCDR, &mut data);
+/// assert!(u32::from_le_bytes(data) > (v as u32));
+/// ```
+pub struct RTC<T: Trigger> {
+    // Counts up from 1 on reset at 1Hz (emulated).
+    counter: Instant,
+
+    // The offset value applied to the counter to get the RTC value.
+    counter_offset: u32,
+
+    // The comparison value used as an alarm (not implemented).
+    // TODO: Implement the match register functionality.
+    // match_value: u32,
+
+    // The interrupt mask.
+    imsc: u32,
+
+    // The raw interrupt value.
+    ris: u32,
+
+    // Used for notifying the driver about events.
+    interrupt_evt: T,
+}
+
+/// Errors encountered while handling RTC operations.
+#[derive(Debug)]
+pub enum Error<E> {
+    /// Failed to trigger interrupt.
+    Trigger(E),
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadWriteOffset(u64),
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadReadOffset(u64),
+}
+
+impl<T: Trigger> RTC<T> {
+    /// Creates a new `AMBA PL031 RTC` instance which uses `trigger` object
+    /// to notify the driver about new events.
+    ///
+    /// # Arguments
+    /// * `trigger` - The Trigger object that will be used to notify the driver
+    ///               about events.
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn new(trigger: T) -> RTC<T> {
+        RTC {
+            // Counts up from 1 on reset at 1Hz (emulated).
+            counter: Instant::now(),
+
+            // The counter_offset is initialized to zero.
+            counter_offset: 0,
+
+            // The alarm value is initialised to zero (not implemented).
+            // TODO: Implement the match register functionality.
+            // match_value: 0,
+
+            // The interrupt mask is initialised as not set.
+            imsc: 0,
+
+            // The raw interrupt is initialised as not asserted.
+            ris: 0,
+
+            interrupt_evt: trigger,
+        }
+    }
+
+    /// Provides a reference to the interrupt event object.
+    pub fn interrupt_evt(&self) -> &T {
+        &self.interrupt_evt
+    }
+
+    fn trigger_interrupt(&mut self) -> Result<(), T::E> {
+        self.interrupt_evt.trigger()
+    }
+
+    fn get_rtc_value(&self) -> u32 {
+        // Add the counter offset to the seconds elapsed since reset.
+        (self.counter.elapsed().as_secs() as u32) + self.counter_offset
+    }
+
+    fn write_imsc(&mut self, val: u32) -> Result<(), T::E> {
+        self.imsc = val & 1;
+        self.trigger_interrupt()?;
+        Ok(())
+    }
+
+    fn write_ris(&mut self) -> Result<(), T::E> {
```

Comment:
I think that the name `write_ris()` is a bit misleading here, since RTCRIS is a read-only register, and this handler is being called as a result of a write to RTCICR.

Reply:
I agree.  As above, I've removed the method and simply update `self.ris` in the `write()` method when `1` is written to `RTCICR`.